Refactor of BitcoinTestFramework main function.

jachiang opened the pull request:
In order to wrap the BitcoinTestFramework class in a user-defined class, which can be started and shutdown, the code in BitcoinTestFramework main has been encapsulated in respective functions. Additionally, test-run specific state in temp-directory and NetworkThread is now reset after shutdown.

_Behavioural changes:_
Potential exceptions raised from any code in setup section of the code are now handled. Previously, only exceptions raised by code beginning [here](https://github.com/jachiang/bitcoin/blob/master-testframework-main-refactor/test/functional/test_framework/test_framework.py#L201) was handled.

_TODO's:_
- [ ] ~~Refactor NetworkThread class in mininode.py, so each BitcoinTestFramework object initialises a different network [thread](https://github.com/bitcoinops/bitcoin/blob/optech-taproot/test/functional/test_framework/mininode.py#L436).~~
- [x] Documentation (markdown file) describing how to run an interactive test with TestWrapper.

jnewbery commented on test/functional/test_framework/mininode.py:
```diff
@@ -480,7 +480,7 @@ def close(self, timeout=10):
         self.network_event_loop.close()
         self.join(timeout)
-
+        NetworkThread.network_event_loop = None # Safe to remove event loop.
```
nit: the convention seems to be that comments usually go the line above in the python test code

jachiang replied:
Fixed in b1f803d03fec4cf4f66cb142ddde705b96aa8205.

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -105,6 +105,17 @@ def __init__(self):
+            self.run_test()
+        except BaseException as exception:
+            e = exception
+        self.shutdown(e = e, exit = True)
```
The second parameter (`exit`) isn't used in this repo. Can you remove it by having `self.shutdown` return the `exit_code` and then `main()` calls `sys.exit()`.

jachiang replied:
Fixed in e163c98169e0405255a2e6acdbb679494b5fa42d.

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -181,33 +196,25 @@ def main(self):
+    def shutdown(self, e=None, exit=False):
+        """Call this method to shutdown the test object and optionally handle an exception."""
+
+        if e != None:
```
I think it makes more logical sense to have the exception handling code in the `main()` function, and then just start `shutdown()` at the `if self.success == TestStatus.FAILED and ...` line

jachiang replied:
I am not sure I fully understand your suggestion. If we move error handling into main, this is code that needs to be replicated in TestWrapper.

jnewbery replied:
why not:

```py
        try:
            self.setup()
            self.run_test()
        except BaseException as e:
            handle_exception(e)
        finally:
            exit_code = self.shutdown()
            sys.exit(exit_code)
```

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -251,7 +258,26 @@ def main(self):
+        if exit:
+            sys.exit(exit_code)
+
+    def handle_exception(self, e):
```
my preference would be to not have this as a separate function from `main()`

jachiang replied:
Addressed in e163c98169e0405255a2e6acdbb679494b5fa42d.

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -440,7 +466,7 @@ def sync_all(self, nodes=None, **kwargs):
     def _start_logging(self):
         # Add logger and logging handlers
-        self.log = logging.getLogger('TestFramework')
+        self.log = logging.getLogger('TestFramework.'+ self.options.tmpdir) # Assign new logger name to prevent temp path reuse.
```
This isn't quite right. Since `logging.shutdown()` is called in `TestFramework.shutdown()`, no more logging should take place (see https://docs.python.org/3/library/logging.html#logging.shutdown).

A better way to do this would be to replace the `logging.shutdown()` call with:

```py
handlers = self.log.handlers
for h in handlers:
    h.flush()
    h.flush()
    self.log.removeHandler(h)
```

That way, when `setup` is called again, the old handlers will have been removed and the new handlers will be added to the `Logging` object.

jachiang replied:
Thanks for this suggestion. Added to e163c98169e0405255a2e6acdbb679494b5fa42.

jnewbery commented:
You should also reword the second commit log to remove reference to `TestWrapper` if you want to get this merged to Bitcoin Core.

jnewbery commented:
> You should also reword the second commit log to remove reference to TestWrapper if you want to get this merged to Bitcoin Core.

Thinking about this some more, perhaps a good way to get this merged to Bitcoin Core would be to include the `TestWrapper` and add some documentation about how to run an interactive test. That's quite a compelling rationale for making these changes.

I'd hold off on that while we're still iterating on `TestWrapper`, but after the workshops let's open a PR against https://github.com/bitcoin/bitcoin.

jachiang commented:
I am adding another TODO for this PR.
- [ ] Refactor NetworkThread class in mininode.py, so each BitcoinTestFramework object initialises a different network [thread](https://github.com/bitcoinops/bitcoin/blob/optech-taproot/test/functional/test_framework/mininode.py#L436).

jnewbery commented:
> Refactor NetworkThread class in mininode.py, so each BitcoinTestFramework object initialises a different network thread.

I don't think we want more than one instance of `BitcoinTestFramework` to exist at a time. @jachiang - do you have a reason for wanting this?

jachiang committed:
Added TestWrapper class.

A BitcoinTestFramework child class which can be imported by an
external user or project. TestWrapper.setup() initiates an underlying
BitcoinTestFramework object with bitcoind subprocesses, rpc interfaces
and test logging. TestWrapper.shutdown() safely tears down the BitcoinTestFramework object.

jachiang reviewed (commented):
> > Refactor NetworkThread class in mininode.py, so each BitcoinTestFramework object initialises a different network thread.
> 
> I don't think we want more than one instance of `BitcoinTestFramework` to exist at a time. @jachiang - do you have a reason for wanting this?

I will omit this from this PR, because it requires limiting one NetworkThread per BitcoinTestFramework, which I am unsure of how to best implement with minimal code changes.

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -251,7 +260,29 @@ def main(self):
-        sys.exit(exit_code)
+
+        for h in list(self.log.handlers):
+            self.log.removeHandler(h)
```
Do these need to be flushed before being removed?

jnewbery commented on test/functional/test_framework/test_framework.py:
```diff
@@ -251,7 +260,29 @@ def main(self):
+
+    def handle_exception(self, e):
+        # Restore traceback from exception object.
+        tb = traceback.format_exception(type(e), e, e.__traceback__)
```
Is this new `traceback` functionality necessary? Can it be separated into a new commit so you're not moving and changing code in the same commit?

jachiang replied:
Not Ideal, I agree. It seems the logger will only capture the traceback if error logging is called directly inside `except` (see https://stackoverflow.com/a/5191885 and https://docs.python.org/3/library/logging.html#logging.error). Since we are invoking error/exception logging elsewhere, we reconstruct the traceback string from the exception object.

jnewbery commented on test/functional/test_framework/test_wrapper.py:
```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2014-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+from test_framework.test_framework import BitcoinTestFramework
```
pep8 import ordering please.

More generally, can you run flake8 over this new file and resolve any style issues?

jnewbery commented on test/functional/test_framework/test_wrapper.py:
```diff
@@ -0,0 +1,104 @@
+
+        def set_test_params(self):
+            # This can be overriden in setup() parameter.
+            self.num_nodes=3
```
is this required? `num_nodes` defaults to 3 in the `setup()` method.

jnewbery commented on test/functional/test_framework/test_wrapper.py:
```diff
@@ -0,0 +1,104 @@
+            bitcoind=abspath(join(__file__ ,"../../../..") +  "/src/bitcoind"),
+            bitcoincli=None,
+            setup_clean_chain=True,
+            num_nodes=3,
```
I suggest you default this to 1

jnewbery commented on test/functional/test_framework/test_wrapper.py:
```diff
@@ -0,0 +1,104 @@
+            pdbonfailure=False,
+            usecli = False,
+            perf = False,
+            randomseed = None):
```
Have you tried changing the method signature to catch `*kwargs` and then set the defaults below, to avoid this huge function signature?

jachiang replied:
Yes, it is much cleaner as you suggested. Thank you!

jnewbery reviewed (commented):
A few comments inline. I also think the first commit comment ("Removal of residual NetworkThread state.") could be expanded a bit to explain why you're making that change.

Once the documentation is added, I think this is ready to open as a PR against bitcoin/bitcoin

jachiang committed:
Added documentation for test_wrapper submodule.

jachiang committed:
Removal of network_event_loop when NetworkThread instance is closed.

The asyncio.new_event_loop() instance is now removed from the NetworkThread class during shutdown. This enables a NetworkThread instance to be restarted after being closed. The current NetworkThread class guards against an existing new_event_loop during initialization.

jachiang committed:
Moved assert num_nodes is set into main().

This allows a BitcoinTestFramework child class to set test parameters
in an overridden setup() rather than in an overridden set_test_params().

jachiang committed:
Added removal of logging handlers.

In order for BitcoinTestFramework to correctly restart after shutdown,
the previous logging handlers need to be removed, or else logging will
continue in the previous temp directory. "Flush" ensures buffers are
emptied, and "close" ensures file handler close logging file.

jachiang committed:
Clear TestNode objects after shutdown.

TestNode objects need to be removed during shutdown, as setup_nodes
does not remove previous TestNode objects from previous test runs during
setup.

jachiang committed:
Refactored BitcoinTestFramework main method into setup and shutdown.

Setup and shutdown code now moved into dedicated methods. Test "success" is
added as a BitcoinTestFramework member, which can be accessed outside of
main.

jachiang commented:
@jnewbery Sorry this took a while, but I've addressed all feedback, reworked commits for easier review and added a documentation file for the TestWrapper class.

I will review again tomorrow and then open a PR against master. 
A very big thank you for your time spent reviewing and improving this!

jnewbery commented:
PR opened against bitcoin/bitcoin here: https://github.com/bitcoin/bitcoin/pull/17288
